text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+For this reason, most of `std::sync::OnceCell` API can not be provided in `core`.
+In the `sync` case, reliably panicking on re-entrant initialization is not trivial.
+For this reason, the implementaion would simply deadlock, with a note that a deadlock might be elevated to panic in the future.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This is a moderately large addition to stdlib, there's a chance we do something wrong.
+  This can be mitigated by piece-wise stabilization (in particular, `Lazy` convenience types are optional) and the fact that API is tested in the crates.io ecosystem via `once_cell` crate.
+
+* The design of `Lazy` type uses default type-parameter as a work-around for the absence of type-inference of statics.
+
+* We use the same name for unsync and sync types, which might be confusing.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why not `Lazy` as a primitive?
+
+On the first look, it may seem like we don't need `OnceCell`, and should only provide `Lazy`.
+The critical drawback of `Lazy` is that it's not always possible to provide the closure at creation time.
+
+This is important for lazy fields:
+
+```rust
+struct Ctx {
+    config_path: PathBuf,
+    config: Lazy<String, ???>,
+}
+
+impl Ctx {
+    pub fn new(config_path: PathBuf) -> Ctx {
+        Ctx {
+            config_path,
+            config: Lazy::new(|| {
+                // We would like to write something like
+                // `fs::read_to_string(&self.config_path)`
+                // here, but we can't have access to `self`
+                ???
+            })
+        }
+    }
+}
+```
+
+Or for singletons, initialized with parameters:
+
+```rust
+use std::{env, io, sync::OnceCell};
+
+#[derive(Debug)]
+pub struct Logger { ... }
+
+static INSTANCE: OnceCell<Logger> = OnceCell::new();
+impl Logger {
+    pub fn global() -> &'static Logger {
+        INSTANCE.get().expect("logger is not initialized")
+    }
+    fn from_cli(args: env::Args) -> Result<Logger, std::io::Error> { ... }
+}
+
+fn main() {
+    let logger = Logger::from_cli(env::args()).unwrap();
+
+    // Note how we use locally-created value for initialization.
+    INSTANCE.set(logger).unwrap();
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies a common use-case.
```

Comment:
To be honest this alternative feels like a useless snowflake, maybe nice for a specialized crate. I think it is better to concentrate on the main insight: `OnceCell` is safe because it gets only written once.

This alternative `OnceFlipCell` that combines a `OnceCell` with an enum distracts from that, and imho the RFC is better off without it.

Reply:
Hmm, relevant discussion: https://old.reddit.com/r/rust/comments/bx75zh/once_cell_021_100_macrofree_lazystatic/eq5qev0/